Add basic authentication

brgeen opened the pull request:
**Update Client Credentials flow for Courses API** [Asana Task](https://app.asana.com/0/1206395725449160/1206174762676768/f)

### Changes

This adds logic for the [Password Grant](https://www.oauth.com/oauth2-servers/access-tokens/password-grant/) flow. Which allows us to send Basic Auth (username/password) credentials to Accounts Next. 

### Test
Reference [Courses API PR](https://github.com/Arux-Software/onsite_courses_api/pull/37)

brgeen committed:
add basic authentication public method

brgeen committed:
remove redirect_uri and add client_secret

goronfreeman commented on lib/arux_app/api/auth.rb:
```diff
@@ -60,6 +60,32 @@ def authorization_url(scope: "public")
+          client_id: client_id,
+          client_secret: client_secret
+        }
+        request = HTTPI::Request.new
+        request.url = "#{self.class.server_uri}/oauth/token"
+        request.body = params
+        request.headers = { 'User-Agent' => USER_AGENT }
+        request.auth.basic(username, password)
```
I looked at the [source](https://github.com/savonrb/httpi/blob/main/lib/httpi/request.rb#L17-L23) for `HTTPI::Request`, and it looks like you can do this:

```suggestion
        request = HTTPI::Request.new(
          url: "#{self.class.server_uri}/oauth/token",
          body: params,
          headers: { "User-Agent" => USER_AGENT }
        )
        request.auth.basic(username, password)
```

That still leaves `request.auth.basic` that you can't pass in. You could also do:

```rb
        request = HTTPI::Request.new.tap do |req|
          req.url = "#{self.class.server_uri}/oauth/token"
          req.body = params
          req.headers = { 'User-Agent' => USER_AGENT }
          req.auth.basic(username, password)
        end
```

This is a super minor thing, but I like to be able to visually see a chunk of stuff that is all working to build up an object as one big thing, not a series of separate lines, if that makes sense. Just makes it a little more readable for me personally.

brgeen replied:
Good call on `#tap` works well and improves readability like you mentioned.

goronfreeman commented on lib/arux_app/api/auth.rb:
```diff
@@ -60,6 +60,32 @@ def authorization_url(scope: "public")
+        request.auth.basic(username, password)
+        response = HTTPI.post(request)
+
+        if !response.error?
+          AccessToken.new(
+            token: JSON.parse(response.body)['access_token'],
+            scope: JSON.parse(response.body)['scope'],
+            auth: self
+          )
+        else
+          raise(API::Error.new(response.code, response.body))
+        end
```
You could do a guard here:

```suggestion
        raise(API::Error.new(response.code, response.body)) if response.error?
        
        AccessToken.new(
          token: JSON.parse(response.body)['access_token'],
          scope: JSON.parse(response.body)['scope'],
          auth: self
        )
```

brgeen replied:
Nice.

goronfreeman commented on lib/arux_app/api/auth.rb:
```diff
@@ -60,6 +60,32 @@ def authorization_url(scope: "public")
+          raise(API::Error.new(response.code, response.body))
+        end
+      end
+
```
```suggestion
```

goronfreeman reviewed (commented):
A couple super small things just from looking at the code. Will test it out tomorrow.

brgeen committed:
build the request up as a block

This improves readability and consolidates the
logic into one chunk.

brgeen committed:
use raise instead of conditional statements

This is using raise as a guard instead of nesting
the return values in conditionals.

brgeen committed:
remove extra line between method signatures

Co-authored-by: Hunter Braun <hunter.braun@gmail.com>
